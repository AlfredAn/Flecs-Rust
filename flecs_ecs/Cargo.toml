[package]
name = "flecs_ecs"
version = "0.0.2"
edition = "2021"
authors = ["Indra de Backere <debackere.indra@gmail.com>"]
description = "Rust API for the C flecs ECS library <https://github.com/SanderMertens/flecs>"
license = ""
keywords = ["ecs", "flecs", "entity", "component", "system"]
categories = ["game-development", "api-bindings"]

[dependencies]
#debug-here = "0.2.2" #temp for development
flecs_ecs_derive = { path = "../flecs_ecs_derive" }
libc = "0.2.148"
seq-macro = "0.3.5" #temp for testing / development
rand = "0.8.5"  #temp for testing / development
#cargo-show-asm = "0.2.21"  #temp for testing / development

[dev-dependencies]
criterion = "0.5.1"
seq-macro = "0.3.5"
rand = "0.8.5" 

[[bench]]
name = "query"
harness = false

[build-dependencies]
bindgen = "0.69.4"
cc = "1.0.83"
regex = "1.9.5"

# ---> Bindgen <--- with bindgen macos generates u32 for the underlying enum of some bindings, 
#which is a bug. It should be generating i32. If you generate from macos, you'll need to revert the
# u32 to i32 in the generated bindings for the enums. More info on the issue here:
#https://github.com/rust-lang/rust-bindgen/issues/1966

[features]
flecs_generate_bindings = [] 
flecs_system = []
flecs_rules = []
flecs_timer = []
flecs_doc = []
flecs_rest = []
flecs_meta = []
flecs_metrics = []
flecs_ecs_asserts = []
flecs_enable_all_asserts = ["flecs_ecs_asserts"]
default = [
    "flecs_enable_all_asserts",
    "flecs_system",
    "flecs_timer",
    "flecs_doc",
    "flecs_rest",
    "flecs_meta",
    "flecs_metrics",
    "flecs_rules",
    #"flecs_generate_bindings"
]

[target.'cfg(any())'.dependencies]
flecs_ecs_derive = { version = "=0.1.0", path = "../flecs_ecs_derive" }