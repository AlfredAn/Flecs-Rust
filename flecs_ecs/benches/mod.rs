use std::time::Duration;

use criterion::{criterion_group, criterion_main, Criterion};
// include!("common.rs") is used so this file can exist and we can manually control the order of the benchmarks

mod add_existing;
mod add_remove;
mod create_delete_entities;
mod entity;
mod event;
mod get;
mod has;
mod hooks;
mod observer;
mod query;
mod relationships;
mod set;

use add_existing::*;
use add_remove::*;
use create_delete_entities::*;
use entity::*;
use event::*;
use get::*;
use has::*;
use hooks::*;
use observer::*;
//use query::*;
use relationships::*;
use set::*;

fn ecs_default_criterion() -> Criterion {
    let mut criterion_config = Criterion::default().configure_from_args();
    criterion_config = criterion_config
        .warm_up_time(Duration::from_millis(500))
        .measurement_time(Duration::from_secs(3))
        .sample_size(50)
        .noise_threshold(0.01)
        .confidence_level(0.95)
        .significance_level(0.05)
        .without_plots();
    criterion_config
}

criterion_main!(
    has,
    get,
    set,
    g_add_remove,
    g_create_delete_entities,
    hooks,
    observers,
    entity,
    event
);

criterion_group!(
    name = has;
    config = ecs_default_criterion();
    targets =
    has_component_not_found,
    has_components_found
);

criterion_group!(
    name = get;
    config = ecs_default_criterion();
    targets =
    get_component_not_found,
    get_components_found,
    get_components_not_found,
    get_mut_components_found,
    get_mut_components_not_found,
    ensure_mut_components_found,
    ensure_mut_components_found_cmd,
    ensure_mut_not_found_and_remove,
    ensure_mut_not_found_and_remove_cmd,
);

criterion_group!(
    name = add_existing;
    config = ecs_default_criterion();
    targets =
    add_existing_operations,
    add_existing_cmd,
);

criterion_group!(
    name = g_add_remove;
    config = ecs_default_criterion();
    targets =
    add_remove,
    add_remove_cmd,
    add_remove_1_tag_to_entity_with_n_components
);

criterion_group!(
    name = set;
    config = ecs_default_criterion();
    targets =
    set_remove,
    set_remove_cmd,
    set_found,
    set_found_cmd
);

criterion_group!(
    name = g_create_delete_entities;
    config = ecs_default_criterion();
    targets =
    create_delete_entities,
    create_delete_entities_cmd,
    create_delete_entities_w_tree
);

criterion_group!(
    name = hooks;
    config = ecs_default_criterion();
    targets =
    add_remove_hooks_components
);

criterion_group!(
    name = relationships;
    config = ecs_default_criterion();
    targets =
    get_relationship_targets,
    override_components_add_remove,
    get_inherited_w_depth,
    change_parent
);

criterion_group!(
    name = observers;
    config = ecs_default_criterion();
    targets =
    observer_create_w_add
);

criterion_group!(
    name = entity;
    config = ecs_default_criterion();
    targets =
    entity_set_name
);

criterion_group!(
    name = event;
    config = ecs_default_criterion();
    targets =
    event_emit,
    event_emit_propagate,
    event_emit_forward,
    event_modified
);
