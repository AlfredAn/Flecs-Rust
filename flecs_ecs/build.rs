#![allow(clippy::all)]
#![allow(warnings)]

#[cfg(feature = "flecs_generate_bindings")]
fn generate_bindings() {
    use std::{env, path::PathBuf};

    #[derive(Debug)]
    struct CommentsCallbacks;

    impl bindgen::callbacks::ParseCallbacks for CommentsCallbacks {
        fn process_comment(&self, comment: &str) -> Option<String> {
            // 1: trimming the comments
            let comment = comment.trim();
            // 2: brackets do not entail intra-links
            let comment = comment.replace("[", "\\[");
            let comment = comment.replace("]", "\\]");

            // ensure all links are padded with < and >
            let url_re = regex::Regex::new(r"(?P<url>https?://[^\s]+)").unwrap();
            let comment = url_re
                .replace_all(comment.as_str(), |caps: &regex::Captures| {
                    format!("<{}>", &caps["url"])
                })
                .into_owned();

            Some(comment)
        }
    }

    let mut bindings = bindgen::Builder::default()
        .header("src/sys/flecs.h")
        // Only keep things that we've allowlisted rather than
        // recursively keeping nested uses around.
        .allowlist_file("src/sys/flecs.h")
        .allowlist_recursively(false)
        // Keep comments and keep all of them, not just doc comments.
        .generate_comments(true)
        .clang_arg("-fparse-all-comments")
        .parse_callbacks(Box::new(CommentsCallbacks))
        .blocklist_item("FLECS_HI_COMPONENT_ID")
        .blocklist_item("ECS_PAIR")
        .blocklist_item("ECS_OVERRIDE")
        .blocklist_item("ECS_TOGGLE")
        .blocklist_item("ECS_AND")
        // We'll use `libc::FILE` instead.
        .blocklist_type("FILE")
        // These have doc comments that trigger doc tests.
        .blocklist_type("ecs_alert_desc_t")
        .blocklist_function("ecs_http_server_http_request")
        .blocklist_function("ecs_log_enable_timedelta")
        // These use va_list
        .blocklist_function("ecs_logv_")
        .blocklist_function("ecs_printv_")
        .blocklist_function("ecs_parser_errorv_")
        .blocklist_function("ecs_strbuf_vappend")
        .blocklist_function("ecs_vasprintf")
        .layout_tests(false)
        .raw_line("#![allow(clippy::all)]")
        .raw_line("#![allow(warnings)]")
        .raw_line("use super::*;")
        .raw_line("use libc::FILE;");

    let bindings = bindings.generate().expect("Unable to generate bindings");

    let crate_root: PathBuf = env::var("CARGO_MANIFEST_DIR").unwrap().into();
    bindings
        .write_to_file(crate_root.join("src/sys/bindings.rs"))
        .unwrap();
}

fn main() {
    // Tell cargo to invalidate the built crate whenever the sources change
    println!("cargo:rerun-if-changed=src/sys/flecs.h");
    println!("cargo:rerun-if-changed=src/sys/flecs.c");
    println!("cargo:rerun-if-changed=build.rs");

    #[cfg(feature = "flecs_generate_bindings")]
    generate_bindings();

    #[cfg(not(feature = "flecs_disable_build_c_library"))]
    // Compile flecs
    cc::Build::new()
        //.compiler("clang")
        //.opt_level(3)
        //.shared_flag(true)
        .warnings(true)
        .extra_warnings(true)
        //.define("NDEBUG", None)
        .file("src/sys/flecs.c")
        .compile("flecs");
}
