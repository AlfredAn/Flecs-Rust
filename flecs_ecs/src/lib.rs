#![allow(dead_code)]

pub mod core;

pub mod addons;

pub mod sys;
