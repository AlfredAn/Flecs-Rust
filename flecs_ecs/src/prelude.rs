pub use crate::addons::module::*;
pub use crate::addons::*;
pub use crate::core::*;
pub use crate::macros::*;
pub use flecs_ecs_sys::EcsComponent;
