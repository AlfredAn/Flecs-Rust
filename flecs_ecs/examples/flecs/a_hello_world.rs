use crate::z_snapshot_test::*;
snapshot_test!();
use flecs_ecs::prelude::*;

#[derive(Debug, Component)]
pub struct Position {
    pub x: f32,
    pub y: f32,
}

#[derive(Debug, Component)]
pub struct Velocity {
    pub x: f32,
    pub y: f32,
}

#[derive(Component)]
pub struct Eats;

#[derive(Component)]
pub struct Apples;

#[test]
fn main() {
    // Create a new world
    let world = World::new();

    //ignore snap in example, it's for snapshot testing
    world.import::<Snap>();

    // Register system
    let _sys = world
        .system::<(&mut Position, &Velocity)>()
        // .each_entity if you want the entity to be added in the parameter list
        .each(|(pos, vel)| {
            pos.x += vel.x;
            pos.y += vel.y;
        });

    // Create an entity with name Bob, add Position and food preference
    let bob = world
        .entity_named(c"Bob")
        .set(Position { x: 0.0, y: 0.0 })
        .set(Velocity { x: 1.0, y: 2.0 })
        .add::<(Eats, Apples)>();

    // Show us what you got
    fprintln!(&world, "{}'s got [{:?}]", bob.name(), bob.archetype());

    // Run systems twice. Usually this function is called once per frame
    world.progress();
    world.progress();

    // - get panics if the component is not present, use try_get for a non-panicking version which does not run the callback.
    // - or use Option to handle the individual component missing.
    bob.get::<&Position>(|pos| {
        // See if Bob has moved (he has)
        fprintln!(&world, "{}'s position: {:?}", bob.name(), pos);
    });

    // Option example
    let has_run = bob.try_get::<Option<&Position>>(|pos| {
        if let Some(pos) = pos {
            // See if Bob has moved (he has)
            fprintln!(&world, "{}'s try_get position: {:?}", bob.name(), pos);
        }
    });

    if has_run {
        fprintln!(
            &world,
            "Bob has a position component, so the try_get callback ran."
        );
    }

    world.get::<&Snap>(|snap| snap.test("hello world".to_string()));

    // Output:
    //  Bob's got [Position, Velocity, (Identifier,Name), (Eats,Apples)]
    //  Bob's position: Position { x: 2.0, y: 4.0 }
    //  Bob's try_get position: Position { x: 2.0, y: 4.0 }
    //  Bob has a position component, so the try_get callback ran.
}
