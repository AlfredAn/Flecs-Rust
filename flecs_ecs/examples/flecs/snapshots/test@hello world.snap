---
source: flecs_ecs/examples/flecs/a_hello_world.rs
expression: self.str
---
- "Bob's got [examples.a_hello_world.Position, examples.a_hello_world.Velocity, (Identifier,Name), (examples.a_hello_world.Eats,examples.a_hello_world.Apples)]"
- "Bob's position: Position { x: 2.0, y: 4.0 }"
- "Bob's try_get position: Position { x: 2.0, y: 4.0 }"
- "Bob has a position component, so the try_get callback ran."
